app/preparing_knights.py

```diff
@@ -0,0 +1,12 @@
+def battle_preparation(knights: dict) -> dict:
+    for knight in knights.values():
+        knight["protection"] = sum(armour["protection"]
+                                   for armour in knight["armour"])
+        knight["power"] += knight["weapon"]["power"]
+
+        if knight["potion"] is not None:
+            for stat in ["power", "protection", "hp"]:
+                if stat in knight["potion"]["effect"]:
+                    knight[stat] += knight["potion"]["effect"][stat]
```

Comment:
I guess better to create potion class, so we can now if there's some effects, if we can modify them, and to ensure there won't be key errors.